0001-qubes-make-fwupdmgr-get-updates-think-it-s-interacti.patch

```diff
@@ -0,0 +1,68 @@
+From 653abf10b5c31cdce1a270b19f5508080766378c Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
+ <marmarek@invisiblethingslab.com>
+Date: Fri, 23 Jun 2023 03:25:28 +0200
+Subject: [PATCH] qubes: make 'fwupdmgr get-updates' think it's interactive
+
+When stdout of fwupdmgr is redirected to a pipe, it assumes it's running
+non-interactive. This makes it not call
+fwupd_client_set_feature_flags(), which results in some (or all) updates not
+being listed.
+
+"Fix" this by using new pty for stdout capture, instead of a pipe.
+---
+ contrib/qubes/src/qubes_fwupdmgr.py | 29 ++++++++++++++++++++++++-----
+ 1 file changed, 24 insertions(+), 5 deletions(-)
+
+diff --git a/contrib/qubes/src/qubes_fwupdmgr.py b/contrib/qubes/src/qubes_fwupdmgr.py
+index 031caf251..ee73af2f4 100755
+--- a/contrib/qubes/src/qubes_fwupdmgr.py
++++ b/contrib/qubes/src/qubes_fwupdmgr.py
+@@ -7,8 +7,10 @@
+ # SPDX-License-Identifier: LGPL-2.1+
+ #
+ 
++import errno
+ import json
+ import os
++import pty
+ import re
+ import shutil
+ import subprocess
+@@ -366,11 +368,28 @@ class QubesFwupdmgr(FwupdHeads, FwupdUpdate, FwupdReceiveUpdates):
+         ):
+             raise ValueError(f"{version} < {self.dmi_version} Downgrade not allowed")
+ 
+-    def _get_dom0_devices(self):
+-        """Gathers information about devices connected in dom0."""
+-        cmd_get_dom0_devices = [FWUPDMGR, "--json", "get-devices"]
+-        p = subprocess.Popen(cmd_get_dom0_devices, stdout=subprocess.PIPE)
+-        self.dom0_devices_info = p.communicate()[0].decode()
++    def _get_dom0_updates(self):
++        """Gathers information about available updates."""
++        cmd_get_dom0_updates = [FWUPDMGR, "--json", "get-updates"]
++        # connect stdout to a pty, otherwise get-updates works in
++        # non-interactive mode and doesn't list all the updates
++        # (based on FWUUPD_FEATURE_* flags)
++        mstdout, sstdout = pty.openpty()
++        p = subprocess.Popen(cmd_get_dom0_updates, stdout=sstdout)
++        os.close(sstdout)
++        data = b""
++        while True:
++            try:
++                new_data = os.read(mstdout, 4096)
++            except OSError as e:
++                if e.errno == errno.EIO:
++                    # process exited
++                    break
++            if not new_data:
++                break
++            data += new_data
```

Comment:
That’s potentially O(N²) in the amount of data sent.  Much better to create a Python list and form the final byte string with `b''.join`.

Reply:
You are likely right, but I don't think it really matters here - we are talking about few kb of data, so this all is <10 calls (usually even <5)